src/material/schematics/ng-update/migrations/m2-theming-v18/migration.ts

```diff
@@ -0,0 +1,282 @@
+  'green-palette',
+  'blue-palette',
+  'yellow-palette',
+  'cyan-palette',
+  'magenta-palette',
+  'orange-palette',
+  'chartreuse-palette',
+  'azure-palette',
+  'violet-palette',
+  'rose-palette',
+];
+
+/** Possible pairs of comment characters in a Sass file. */
+const COMMENT_PAIRS = new Map<string, string>([
+  ['/*', '*/'],
+  ['//', '\n'],
+]);
+
+/** Prefix for the placeholder that will be used to escape comments. */
+const COMMENT_PLACEHOLDER_START = '__<<ngM2ThemingMigrationEscapedComment';
+
+/** Suffix for the comment escape placeholder. */
+const COMMENT_PLACEHOLDER_END = '>>__';
+
+/** Replaces all usages of renamed M2 theming APIs in a file. */
+export function migrateM2ThemingApiUsages(fileContent: string): string {
+  // Strip out comments, so they don't confuse our migration.
+  let {content, placeholders} = escapeComments(fileContent);
+  const materialNamespaces = getNamespaces('@angular/material', content);
+  const experimentalNamespaces = getNamespaces('@angular/material-experimental', content);
+
+  // Migrate the APIs whose names were prefixed with `m2-`.
+  for (const namespace of materialNamespaces) {
+    for (const name of RENAMED_FUNCTIONS) {
+      content = migrateFunction(content, namespace, name, namespace, 'm2-' + name);
+    }
+
+    for (const name of RENAMED_VARIABLES) {
+      content = migrateVariable(content, namespace, name, namespace, 'm2-' + name);
+    }
+  }
+
+  // Migrate themes that were using M3 while it was still in experimental.
+  if (experimentalNamespaces.length > 0) {
+    const preExperimentalContent = content;
+    const stableNamespace = materialNamespaces.length === 0 ? 'mat' : materialNamespaces[0];
+
+    for (const namespace of experimentalNamespaces) {
+      // The only mixin that was renamed was the backwards-compatibility one.
+      content = migrateMixin(
+        content,
+        namespace,
+        'color-variants-back-compat',
+        stableNamespace,
+        'color-variants-backwards-compatibility',
+      );
+
+      // M3 functions weren't prefixed with anything
+      // so they just move over to the new namespace.
+      for (const name of M3_FUNCTIONS) {
+        content = migrateFunction(content, namespace, name, stableNamespace, name);
+      }
+
+      // Variables were all prefixed with `m3-` which needs to be stripped.
+      for (const name of M3_VARIABLES) {
+        content = migrateVariable(content, namespace, 'm3-' + name, stableNamespace, name);
+      }
+    }
+
+    // If experimental is imported, but Material isn't, insert a new import at the top.
+    // This should be rare since `@angular/material` was still required for the theme.
+    if (materialNamespaces.length === 0 && content !== preExperimentalContent) {
+      content = `@use '@angular/material' as ${stableNamespace};\n` + content;
+    }
+  }
+
+  return restoreComments(content, placeholders);
+}
+
+/** Renames all usages of a Sass function in a file. */
+function migrateFunction(
```

Comment:
a lot of these helper functions might be useful to be in some shared place for any future ng-update schematics. Although I guess it's unlikely we would need these in another context soon.

Reply:
Agreed, I actually copied most of these from a migration that we deleted a few versions ago.